manager/orchestrator/task.go

```diff
@@ -65,3 +75,119 @@ func IsTaskDirty(s *api.Service, t *api.Task) bool {
 	return !reflect.DeepEqual(s.Spec.Task, t.Spec) ||
 		(t.Endpoint != nil && !reflect.DeepEqual(s.Spec.Endpoint, t.Endpoint.Spec))
 }
+
+// InvalidNode is true if the node is nil, down, or Drain
+func InvalidNode(n *api.Node) bool {
+	return n == nil ||
+		n.Status.State == api.NodeStatus_DOWN ||
+		n.Spec.Availability == api.NodeAvailabilityDrain
+}
+
+type isRelatedService func(*api.Service) bool
+
+// InitTasks fix tasks in store before orchestrator run. Previous manager might haven't finished processing their updates.
+func InitTasks(ctx context.Context, s *store.MemoryStore, readTx store.ReadTx, relatedService isRelatedService, restart restarter) error {
+	_, err := s.Batch(func(batch *store.Batch) error {
+		tasks, err := store.FindTasks(readTx, store.All)
+		if err != nil {
+			return err
+		}
+		for _, t := range tasks {
+			if t.ServiceID == "" {
+				continue
+			}
+
+			// TODO(aluzzardi): We should NOT retrieve the service here.
+			service := store.GetService(readTx, t.ServiceID)
+			if service == nil {
+				// Service was deleted
+				err := batch.Update(func(tx store.Tx) error {
+					return store.DeleteTask(tx, t.ID)
+				})
+				if err != nil {
+					log.G(ctx).WithError(err).Error("failed to set task desired state to dead")
+				}
+				continue
+			}
+			// only handle tasks belong to related service type
+			if !relatedService(service) {
+				continue
+			}
+			// if the task shouldn't be run on the specified node, set the task to shutdown
+			// so orchestrator can work on it
+			if t.NodeID != "" {
+				n := store.GetNode(readTx, t.NodeID)
+				if InvalidNode(n) && t.DesiredState <= api.TaskStateRunning {
+					t.DesiredState = api.TaskStateShutdown
+					err = batch.Update(func(tx store.Tx) error {
+						return store.UpdateTask(tx, t)
+					})
```

Comment:
I'm not sure this will do the right thing. In the replicated case, the task may need to be restarted, but I believe that if the orchestrator sees a task move its desired state to `TaskStateShutdown`, it will assume it has already been restarted.